src/ray/raylet/main.cc

```diff
@@ -202,10 +121,112 @@ int main(int argc, char *argv[]) {
   }
   RAY_CHECK_OK(gcs_client->Connect(main_service));
 
-  std::unique_ptr<ray::raylet::Raylet> server(new ray::raylet::Raylet(
-      main_service, raylet_socket_name, node_ip_address, redis_address, redis_port,
-      redis_password, node_manager_config, object_manager_config, gcs_client));
-  server->Start();
+  if (head_node) {
+    // Parse the configuration list.
+    std::istringstream config_string(config_list);
+    std::string config_name;
+    std::string config_value;
+
+    while (std::getline(config_string, config_name, ',')) {
+      RAY_CHECK(std::getline(config_string, config_value, ','));
+      // TODO(rkn): The line below could throw an exception. What should we do about this?
+      raylet_config[config_name] = config_value;
+    }
+    RAY_CHECK_OK(gcs_client->Nodes().AsyncSetInternalConfig(raylet_config));
+  }
+
+  std::unique_ptr<ray::raylet::Raylet> server(nullptr);
+
+  RAY_CHECK_OK(gcs_client->Nodes().AsyncGetInternalConfig([&](const std::unordered_map<
+                                                              std::string, std::string>
+                                                                  stored_raylet_config) {
```

Comment:
could there be a race condition where `AsyncGetInternalConfig` returns faster than `AsyncSetInternalConfig`? in that case the head node raylet will not be initialized correctly.

Reply:
Another issue is that now the other raylets implicitly depends on the start of the "head" raylet. We can easily have configuration mismatch happening like:
```
1. user starts head node
2. user adds child node
2. gcs started
3. child node's raylet started, talked to gcs, saw empty config
4. head node's raylet started, talked to gcs, write non-empty config
```